tests/edalize_common.py

```diff
@@ -80,28 +80,30 @@ def setup_backend(paramtypes, name, tool, tool_options, use_vpi=False):
-    {'name' : 'pcf_file.pcf' , 'file_type' : 'PCF'},
-    {'name' : 'ucf_file.ucf' , 'file_type' : 'UCF'},
-    {'name' : 'user_file'    , 'file_type' : 'user'},
-    {'name' : 'tcl_file.tcl' , 'file_type' : 'tclSource'},
-    {'name' : 'vlog_file.v'  , 'file_type' : 'verilogSource'},
-    {'name' : 'vlog05_file.v', 'file_type' : 'verilogSource-2005'},
-    {'name' : 'vlog_incfile' , 'file_type' : 'verilogSource', 'is_include_file' : True},
-    {'name' : 'vhdl_file.vhd', 'file_type' : 'vhdlSource'},
-    {'name' : 'vhdl_lfile'   , 'file_type' : 'vhdlSource', 'logical_name' : 'libx'},
-    {'name' : 'vhdl2008_file', 'file_type' : 'vhdlSource-2008'},
-    {'name' : 'xci_file.xci' , 'file_type' : 'xci'},
-    {'name' : 'xdc_file.xdc' , 'file_type' : 'xdc'},
-    {'name' : 'bootrom.mem'  , 'file_type' : 'mem'},
-    {'name' : 'c_file.c'     , 'file_type' : 'cSource'},
-    {'name' : 'cpp_file.cpp' , 'file_type' : 'cppSource'},
-    {'name' : 'c_header.h'   , 'file_type' : 'cSource', 'is_include_file' : True},
-    {'name' : 'c_header.h'   , 'file_type' : 'cppSource', 'is_include_file' : True},
+    {'name' : 'qip_file.qip'               , 'file_type' : 'QIP'},
+    {'name' : 'qsys_file'                  , 'file_type' : 'QSYS'},
+    {'name' : 'sdc_file'                   , 'file_type' : 'SDC'},
+    {'name' : 'bmm_file'                   , 'file_type' : 'BMM'},
+    {'name' : 'sv_file.sv'                 , 'file_type' : 'systemVerilogSource'},
+    {'name' : 'pcf_file.pcf'               , 'file_type' : 'PCF'},
+    {'name' : 'ucf_file.ucf'               , 'file_type' : 'UCF'},
+    {'name' : 'user_file'                  , 'file_type' : 'user'},
+    {'name' : 'tcl_file.tcl'               , 'file_type' : 'tclSource'},
+    {'name' : 'vlog_file.v'                , 'file_type' : 'verilogSource'},
+    {'name' : 'vlog05_file.v'              , 'file_type' : 'verilogSource-2005'},
+    {'name' : 'vlog_incfile'               , 'file_type' : 'verilogSource', 'is_include_file' : True},
+    {'name' : 'vhdl_file.vhd'              , 'file_type' : 'vhdlSource'},
+    {'name' : 'vhdl_lfile'                 , 'file_type' : 'vhdlSource', 'logical_name' : 'libx'},
+    {'name' : 'vhdl2008_file'              , 'file_type' : 'vhdlSource-2008'},
+    {'name' : 'xci_file.xci'               , 'file_type' : 'xci'},
+    {'name' : 'xdc_file.xdc'               , 'file_type' : 'xdc'},
+    {'name' : 'bootrom.mem'                , 'file_type' : 'mem'},
+    {'name' : 'c_file.c'                   , 'file_type' : 'cSource'},
+    {'name' : 'cpp_file.cpp'               , 'file_type' : 'cppSource'},
+    {'name' : 'c_header.h'                 , 'file_type' : 'cSource', 'is_include_file' : True},
+    {'name' : 'c_header.h'                 , 'file_type' : 'cppSource', 'is_include_file': True},
+    {'name' : 'python_dir/python_file.py'  , 'file_type' : 'pythonSource'},
+    {'name' : 'python_dir2/python_file2.py', 'file_type' : 'pythonSource'},
```

Comment:
What's the effect of adding these here? Is this just for `PYTHON_PATH` population? Or does it also use it to track stale dependencies?

Not all python files should be added to the path, for instance `model/some_helper.py` might be intended to be imported as `import model.some_helper`.

Reply:
Thanks for the comments.

The intention of having the python files in this module is to test that the PYTHONPATH is correctly populated. I'd considered the use case that you suggest and thought that it does not actually break that. (Maybe I'm wrong?)

I'd thought of a couple of ways of effecting the construction of a PYTHONPATH. The solution needs to work when FuseSoC cores provide a hierarchy of models as well as a hierarchy of RTL (example here https://github.com/gchqdev1729/fusesoc-cores/blob/master/gchqdev1729-add_and_invert.core). The two solutions I came up with were passing path components (i.e. directories) or passing the python files as here.

I think passing path components breaks down in a couple of ways. Firstly, FuseSoC explicitly doesn't allow directories in its file list just now. Secondly, it does not ensure that the files are copied from the provider into the FuseSoC workspace where Edalize is going to work on them, so it's unclear how they would be made available at runtime. 

In our own projects, we do have some external dependencies, both open source and proprietary and they would be imported exactly as you suggest (i.e by `import mypackage.mymodule`).

Hope that makes sense...

Reply:
In your example file, how am I suppose to import `add_and_invert/model/model_add_and_invert.py`?

1. `import add_and_invert.model.model_add_and_invert` &rarr; `.` should be in `sys.path`
2. `import model.model_add_and_invert` &rarr; `./add_and_invert` should be in `sys.path`
3. `import model_add_and_invert` &rarr; `./add_and_invert/model` should be in `sys.path`

Putting all three on the path is a very bad idea, as then you have multiple distinct copies of the same module depending on which path you import.

Reply:
OK, I understand. In our code, we have historically used option 3, so what I'm doing in my branch works, but I can see the problem.

I'll have a think to see if there is another solution. I think I need to add all the Python files to the FuseSoC core files to ensure that all the required files are copied to the workspace and made accessible in the file list passed to Edalize. Then I need to construct the PYTHONPATH in some way. I suppose there might be a way to use file attributes to add an ancestor directory other than the direct parent. Beyond that, unless I've missed something, I think it will require wider changes to FuseSoC.

Reply:
You're probably right that this is an issue with FuseSoC itself, and so probably isn't necessary to address immediately - although a comment drawing attention to the shortcoming would be nice.

What happens if you omit some files from FuseSoC? You might be able to list only a root `__init__.py` or `__init__.pyi`, and use that to construct the PYTHON_PATH.

Reply:
Sadly, this wouldn't work as it is because __init__.py would get copied, but not the modules. FuseSoC passes the path of the copied files to Edalize, not where they were copied from, so the module files themselves wouldn't be in the path as constructed.

Reply:
I will document this, as suggested though.

Reply:
Perhaps the python path should be set via some other means then, rather than inferring it from the filenames.

Reply:
I'm pretty sure that the only thing a FuseSoC core inherits from its dependencies is the list of files. For my hierarchical model use case, I need the info to propagate up, so I think this is the only way without wider changes. (I'm a relative novice when it comes to FuseSoC though, so I might be proved wrong!)